src/freenet/crypt/SHA256.java

```diff
@@ -101,9 +87,9 @@ public static MessageDigest getMessageDigest() {
 		try {
 			MessageDigest md = null;
 			synchronized(digests) {
-				int x = digests.size();
-				if(x == 0) md = null;
-				else md = digests.remove(x-1);
+				while (digests.size() > 0 && md == null) {
+					md = digests.remove(0).get();
+				}
```

Comment:
This change increases the worst-case complexity from O(n) to O(n²). You seem to be emulating queue behaviour here to reuse the least recently used digest, why not use a `Queue` instead?